source/include/FreeRTOS_IP.h

```diff
@@ -98,22 +98,30 @@ extern uint32_t ulApplicationGetNextSequenceNumber( uint32_t ulSourceAddress,
  * pointer back to the network buffer.  Should be a multiple of 8 to ensure 8 byte
  * alignment is maintained on architectures that require it.
  *
- * In order to get a 32-bit alignment of network packets, an offset of 2 bytes
- * would be desirable, as defined by ipconfigPACKET_FILLER_SIZE.  So the malloc'd
+ * In order to get a 32-bit or 64-bit alignment of network packets, an offset of 2 bytes
+ * is ideal as defined by ipconfigPACKET_FILLER_SIZE. So the malloc'd
  * buffer will have the following contents:
- *  uint32_t pointer;   // word-aligned
- *  uchar_8 filler[6];
- *  << ETH-header >>    // half-word-aligned
- *  uchar_8 dest[6];    // start of pucEthernetBuffer
- *  uchar_8 dest[6];
- *  uchar16_t type;
- *  << IP-header >>     // word-aligned
+ * +---------+-----------+---------+
+ * | Offset  | Alignment | Length  |
+ * | 32 | 64 | 32  | 64  | 32 | 64 |
+ * |----|----|-----|-----|----|----|
+ * | 0  | 0  | 4   | 8   | 4  | 8  | uchar_8 * pointer;
+ * | 4  | 8  | 4   | 8   | 2  | 6  | uchar_8   filler[];   // ipconfigPACKET_FILLER_SIZE
+ * | 6  | 14 | 4+2 | 8+6 | 7  | 7  | uchar_8   preamble[7] // Beginning of Ethernet header
```

Comment:
I'm not sure if preamble should be word aligned or the ip-header.

Reply:
I think the IP header should be word aligned. The ethernet header is anyways described as an array of uint8_t. It's alignment should not matter I think.

@htibosch care to comment/clarify?

Reply:
@paulbartell wrote:

> I'm not sure if the ethernet preamble should be word aligned or the ip-header.

The Ethernet preamble is used by the hardware layer to establish the bit-speed of the connection. The first byte that is sent to the application (through MDIO or SPI) is the target hardware address. DMA understands the +2 offset from an aligned memory address.

Thank you Paul for the complete overview that compares 32- with 64-bits alignment. I made some small changes:

~~~
/*
 * +---------+-----------+---------+
 * | Offset  | Alignment | Length  |
 * | 32 | 64 | 32  | 64  | 32 | 64 |
 * |----|----|-----|-----|----|----|
 * | 0  | 0  | 4   | 8   | 4  | 8  | uchar_8 * pointer;     // Points to the 'NetworkBufferDescriptor_t'.
 * | 4  | 8  | 4   | 8   | 2  | 2  | uchar_8   filler[];    // To give the +2 byte offset.
 * | 6  | 10 | 4+2 | 8+2 | 6  | 6  | uchar_8   dest_mac[6]; // Destination address.
 * | 12 | 16 | 4   | 8   | 6  | 6  | uchar_8   src_mac[6];  // Source address.
 * | 18 | 22 | 4+2 | 8+6 | 2  | 2  | uchar16_t ethertype;
 * | 20 | 24 | 4   | 8   | ~  | ~  | << IP-header >>        // word-aligned, either 4 or 8 bytes.
 *  uint8_t ucVersionHeaderLength;
 *  etc
 */
~~~

I removed 'preamble' and 'delimiter' from the list.

@AniruddhaKanhere wrote:

> I think the IP header should be word aligned. The ethernet header is anyways described as an array of uint8_t. It's alignment should not matter I think.

That is right, we want the IP-header be aligned at least at 32-bits. There are no 64-bit fields in the library.
If a field is declared as `uint32_t`, it is assumed that the field is 32-bit aligned within the network buffer.
Otherwise a character array is used, as in the ARP header:
~~~
MACAddress_t xSenderHardwareAddress;  /**< Address of the sender     8 +  6 = 14 */
uint8_t ucSenderProtocolAddress[ 4 ]; /**< IP-address of sender     14 +  4 = 18  */
MACAddress_t xTargetHardwareAddress;  /**< Address of the receiver  18 +  6 = 24  */
~~~